wp/lib/bap_wp/src/z3_utils.ml

```diff
@@ -100,6 +100,113 @@ let get_z3_name (map : Constr.z3_expr Var.Map.t) (name : string) (fmt : Var.t ->
+let synonym name ret body = define_fun name [] ret body
+let define_sort (name : string) s = Sexp.List [
+ Sexp.Atom "define-sort"; Sexp.Atom name; Sexp.List []; s]
+let sexp_of_sort (s : Z3.Sort.sort) : Sexp.t = Sexp.of_string (Z3.Sort.to_string s)
+
+(* [make_arg_synonyms] constructs smtlib formula for the arguments of a subroutine.
+   This make synonyms that for high level c-like names ot low level locations *)
+let make_arg_synonyms
+  ?postfix:(postfix = "")
+  (sub : Sub.t)
+  (init_var_map : Expr.expr Var.Map.t)
+  (var_map : Expr.expr Var.Map.t) : Sexp.t list =
+  let args = Term.enum arg_t sub |> Seq.to_list in
+  List.concat_map args ~f:(fun arg ->
+    let cname = Arg.lhs arg in
+    let syn prefix z3_reg =
+      let sort = Sexp.of_string (Z3.Sort.to_string (Z3.Expr.get_sort z3_reg)) in
+      synonym (prefix ^ Var.name cname ^ postfix) sort
+      (Sexp.Atom  (Expr.to_string z3_reg))
+    in
+    let reg = Arg.rhs arg in
+    match reg with
+    | Var reg -> begin
+         match Arg.intent arg with
+           | Some In   -> [syn "init_" (Var.Map.find_exn init_var_map reg)]
+           | Some Out  -> [syn "" (Var.Map.find_exn var_map reg)]
+           | Some Both -> [syn "init_" (Var.Map.find_exn init_var_map reg);
+                           syn "" (Var.Map.find_exn var_map reg)]
+           | None -> []
+    end
+    | _ -> [] (* We don't make synonym unless c variable is just held in a bap variable *)
+    )
+
+(** [args_prelude_compare] builds smtlib synonyms in a comparison for the arguments of
+    two compared subroutines. *)
+let args_prelude_compare
+  (main_sub1 : Sub.t)
+  (main_sub2 : Sub.t)
+  (env1 : Env.t)
+  (env2 : Env.t) : Sexp.t list =
+  let var_map1 = Env.get_var_map env1 in
+  let var_map2 = Env.get_var_map env2 in
+  let init_var_map1 = Env.get_init_var_map env1 in
+  let init_var_map2 = Env.get_init_var_map env2 in
+  let syn1 = make_arg_synonyms ~postfix:"_orig" main_sub1 init_var_map1 var_map1 in
+  let syn2 = make_arg_synonyms ~postfix:"_mod" main_sub2 init_var_map2 var_map2 in
+  syn1 @ syn2
+
+
+
+(** [def_pointer_sort] constructs an smtlib sort synonym of pointer sort.
+    Typically it will be `(define-sort pointer () BV64)` *)
+let def_pointer_sort env =
+  let mem = Env.get_mem env in
+  let var_map = Env.get_var_map env in
+  let z3_mem = Var.Map.find_exn var_map mem in
+  let pointer_sort = Z3.Z3Array.get_domain (Expr.get_sort z3_mem) in
+  define_sort "pointer" (sexp_of_sort pointer_sort)
+
+(** [def_mem_sort] constructs an smtlib sort synonym of the memory array.
+    Typically it will be `(define-sort memsort () (Array (BV64 BV8))` *)
+let def_mem_sort env =
+    let mem = Env.get_mem env in
+    let var_map = Env.get_var_map env in
+    let z3_mem = Var.Map.find_exn var_map mem in
+    define_sort "memsort" (sexp_of_sort (Z3.Expr.get_sort z3_mem))
+
+let compare_prelude_smtlib2  main_sub1 main_sub2 (env1 : Env.t) (env2 : Env.t) : string =
+ let std_prelude = "
```

Comment:
I wonder if we should have some sort of file that contains the prelude that we read in over here rather than having a hard-coded string.

Reply:
I understand the impule, but I'd rather not. A large amount of the string is programmatically generated and going to an external file adds a lot of complexity in terms of finding the file in question from wherever the wp is invoked. I think going to an external file was a mistake we made in vibes.

Reply:
I stand by the vibes decision though.

Reply:
In this case, I feel like the string is short enough to allow minor sins.

Reply:
We added an unnecessary and weird installation step that can easily go wrong or get stale just to avoid recompiling when we as developers edit the string. No user is ever going to edit this string and they can achieve the same functionality by writing `define-fun` in their own smtlib. There is the possibility of adding another command line feature of `--wp-user-lib=mylib.smt2` to avoid duplication between the pre and post condition. I'm not entirely sure we have to have both pre and post condition flags anyway, since this can be achieved via `=>`

Reply:
How so?

Reply:
`(=>   (preconds referring to init_ variables  )  (post-conds  referring to both init and non-init)`. I feel like this is how I've seen Chris write specs in vibes

Reply:
True. I guess we have some redundancy between `init_var` and the `pre` flag.